Fix SPDY proxy authentication with percent escaped characters

3point2 opened the pull request:
Percent escaped characters in the proxy URL are now correctly used unescaped when constructing the proxy auth header for SPDY connections.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
kubectl exec fails when using a proxy that requires authentication and the user's credentials include percent escaped characters. see kubernetes/kubectl#791 for details.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes kubernetes/kubectl#791

#### Special notes for your reviewer:
None

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

leilajal commented:
/assign @seans3 
/triage accepted

3point2 commented:
/priority P2
(carried over from kubernetes/kubectl#791)

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -184,12 +184,15 @@ func (s *SpdyRoundTripper) dialWithHttpProxy(req *http.Request, proxyURL *url.UR
 	if err != nil && err != httputil.ErrPersistEOF {
 		return nil, err
 	}
+	if response != nil && response.StatusCode != http.StatusOK {
```
how is this related?

aojea replied:
separate commits are better in this case , imagine we had to revert because of X reasons, we lost this change that is useful

3point2 replied:
:+1:  done

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -401,15 +411,26 @@ func TestRoundTripAndNewConnection(t *testing.T) {
 			if testCase.proxyAuth != nil {
-				encodedCredentials := base64.StdEncoding.EncodeToString([]byte(testCase.proxyAuth.String()))
-				expectedProxyAuth = "Basic " + encodedCredentials
+				expectedPassword, _ := testCase.proxyAuth.Password()
```
don't you have to check if the password is set?

aojea replied:
I didn't look too much , but it will be nice if we can reuse some of the golang std library methods instead of this manual parsing https://pkg.go.dev/net/http#Request.BasicAuth

aojea replied:
I was curious about the password because in the code above you check that is not empty to generate the credentials

> > 	proxyAuthUser := proxyURL.User.Username()
> 	proxyAuthPassword, proxyAuthPasswordSet := proxyURL.User.Password()
> 	credentials := proxyAuthUser
> 	if proxyAuthPasswordSet {
> 		credentials += ":" + proxyAuthPassword
> 	}

This is what we should do  https://cs.opensource.google/go/go/+/master:src/net/http/transport.go;l=844-856;drc=c81c0279827545b20f81797bb263ae696d9c235f 

in our `func (s *SpdyRoundTripper) proxyAuth(proxyURL *url.URL) string {`

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -294,7 +297,12 @@ func (s *SpdyRoundTripper) proxyAuth(proxyURL *url.URL) string {
 		return ""
 	}
-	credentials := proxyURL.User.String()
+	proxyAuthUser := proxyURL.User.Username()
+	proxyAuthPassword, proxyAuthPasswordSet := proxyURL.User.Password()
+	credentials := proxyAuthUser
+	if proxyAuthPasswordSet {
+		credentials += ":" + proxyAuthPassword
+	}
```
I think that we should do the same as the std library https://cs.opensource.google/go/go/+/refs/tags/go1.19.2:src/net/http/transport.go;l=846;drc=e822b1e26e20ef1c76672c0b77b0fd8a97a1fe84
 , something like 

```
		username := u.Username()
		password, _ := u.Password()
	credentials := fmt.Sprintf("%s:%s", username, password)
```

aojea commented:
/ok-to-test

it lgtm, just some comments to split into 2 different commits and refactor to match the std library logic

3point2 committed:
Improve error message when proxy connection fails

If the proxy doesn't respond to our CONNECT request with a 2XX status
code, we now display the proxy's response rather than the confusing
error "tls: first record does not look like a TLS handshake"

3point2 committed:
Fix SPDY proxy authentication with special chars

The username and password sent in the Proxy-Authorization header are not
supposed to be percent escaped prior to being base64 encoded.

3point2 reviewed (commented):
I've reworked the construction of the proxy auth header to match the [standard library code](https://cs.opensource.google/go/go/+/master:src/net/http/transport.go;l=844-856;drc=c81c0279827545b20f81797bb263ae696d9c235f) that you linked to and split into two commits @aojea, let me know what you think.

aojea commented:
/lgtm
/assign @liggitt

Jordan, for context https://github.com/kubernetes/kubernetes/pull/113283#discussion_r1011005114 , this replaces the auth logic to match the std lib

3point2 commented:
@liggitt just checking in as it's been a week - do you think you'll be able to find a bit of time to look at this PR please?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper.go:
```diff
@@ -184,12 +184,15 @@ func (s *SpdyRoundTripper) dialWithHttpProxy(req *http.Request, proxyURL *url.UR
 	if err != nil && err != httputil.ErrPersistEOF {
 		return nil, err
 	}
+	if response != nil && response.StatusCode != http.StatusOK {
```
limiting this to exactly 200 does not seem correct... change this to accept any 2xx status code, per https://www.rfc-editor.org/rfc/rfc7231#section-4.3.6:

>   Any 2xx (Successful) response indicates that the sender (and all
>   inbound proxies) will switch to tunnel mode immediately after the
>   blank line that concludes the successful response's header section;
>   data received after that blank line is from the server identified by
>   the request-target.  Any response other than a successful response
>   indicates that the tunnel has not yet been formed and that the
>   connection remains governed by HTTP.

3point2 replied:
you're right, thanks for catching that

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -400,16 +410,27 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+				username := testCase.proxyAuth.Username()
+				password, _ := testCase.proxyAuth.Password()
+				auth = username + ":" + password
+				authHeader = "Basic " + base64.StdEncoding.EncodeToString([]byte(auth))
```
keep the name of this `expectedProxyAuth` to make it clear which is the data in the testcase and which is the data we're expecting

both the old test code and new test code share a flaw of essentially duplicating the implementation, so if the implementation is wrong, our copy in the test will be wrong too, and the test will still pass

can we improve this to actually extract the username and password from the header using a stdlib method and see if the username and password match the testcase data? (example at https://go.dev/play/p/psQsaWVA4ZG)

3point2 replied:
it's now expectedUsername and expectedPassword, thanks for the suggestion. although it doesn't get rid of all the duplication of logic, at least the base64 decoding is now being handled by stdlib.

liggitt reviewed (commented):
Sorry, this got stuck in the crush of code freeze reviews. Did a sweep

3point2 commented on staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/roundtripper_test.go:
```diff
@@ -291,6 +291,16 @@ func TestRoundTripAndNewConnection(t *testing.T) {
+		"proxied valid https, proxy auth with special char in password -> valid https": {
+			serverFunc:             httpsServerValidHostname(t),
+			proxyServerFunc:        httpsServerValidHostname(t),
+			proxyAuth:              url.UserPassword("proxyuser", "proxypasswd%"),
```
done

3point2 reviewed (commented):
> Sorry, this got stuck in the crush of code freeze reviews

No problem, thanks for replying and reviewing so quickly! I've tried to address your comments, please let me know what you think.

3point2 commented:
/retest

liggitt commented:
/lgtm
/approve

I don't feel strongly about this being included in the 1.26 milestone, but I'd be supportive of a cherry-pick to release branches. Not sure how the release team would want a low risk bugfix PR like that dispositioned.

3point2 commented:
Thanks for the review!

Is there anything else I need to do to help this get merged, or is it just a matter of waiting until the 1.26 freeze ends?

liggitt commented:
you're good, just waiting for master to open for 1.27

dims commented:
see discussion above, punting to post code freeze.

/milestone v1.27
